### PR TITLE
add support for taproot btc

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Resolves backend aspect of #4449

### CODE REVIEW

File: rotkehlchen/chain/bitcoin/utils.py [hunk 1]
```diff
@@ -107,21 +188,91 @@ def pubkey_to_p2sh_p2wpkh_address(data: bytes) -> BTCAddress:
 
 def pubkey_to_bech32_address(data: bytes, witver: int) -> BTCAddress:
     """
-    Bitcoin pubkey to bech32 address
```
Comment (nebolax) [inline rotkehlchen/chain/bitcoin/utils.py:110]: I would add in the docstring what is `witver`

File: rotkehlchen/chain/bitcoin/utils.py [hunk 2]
```diff
@@ -107,21 +188,91 @@ def pubkey_to_p2sh_p2wpkh_address(data: bytes) -> BTCAddress:
...
+
+def _bech32_encode(data: List[int], spec: BitcoinSegwitEncoding) -> str:
+    combined = data + _bech32_create_checksum(data, spec)
+    return 'bc1' + ''.join([bech32.CHARSET[d] for d in combined])
+
 
-    return BTCAddress(result)
+def _bech32_create_checksum(data: List[int], spec: BitcoinSegwitEncoding) -> list:
+    values = bech32.bech32_hrp_expand('bc') + data
+    const = BECH32M_CONST if spec == BitcoinSegwitEncoding.BECH32M else 1
```
Comment (nebolax) [inline rotkehlchen/chain/bitcoin/utils.py:273]: If I understand correctly, the only difference between this method and `bech32.create_checksum` is the `BECH32M_CONST`. So I'm just curious, what is it?

File: rotkehlchen/tests/api/test_blockchain.py [hunk 3]
```diff
@@ -587,6 +587,18 @@ def test_add_blockchain_accounts(
...
+    response = requests.put(api_url_for(
+        rotkehlchen_api_server,
+        'blockchainsaccountsresource',
+        blockchain='BTC',
+    ), json={'accounts': [
+        {'address': 'bc1pqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqszqgpqyqsyjer9e'},
+    ]})
+
+    result = assert_proper_response_with_result(response)
+    assert result['per_account']['BTC'] is not None
```
Comment (nebolax) [inline rotkehlchen/tests/api/test_blockchain.py:600]: I think this would need a rebase since we have recently removed balances return from accounts addition

File: rotkehlchen/chain/bitcoin/utils.py [hunk 4]
```diff
@@ -107,21 +188,91 @@ def pubkey_to_p2sh_p2wpkh_address(data: bytes) -> BTCAddress:
...
     Source:
     https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#witness-program
-    https://github.com/mcdallas/cryptotools/blob/master/btctools/address.py
+    https://github.com/bitcoin/bips/blob/master/bip-0086.mediawiki
 
     May raise:
     - EncodingError if address could not be derived from public key
     """
-    witprog = hash160(data)
-    result = bech32.encode('bc', witver, witprog)
```
Comment (nebolax) [inline rotkehlchen/chain/bitcoin/utils.py:120]: Hm... looks a bit suspicious that previously this function was already supporting different `witver` values. So what is the difference between new logic and passing `witver=1` to this function in the past?

File: rotkehlchen/chain/bitcoin/utils.py [hunk 5]
```diff
@@ -21,18 +30,90 @@ class OpCodes:
...
+        is_valid_base58_address(value) or
+        is_valid_bech32_address(value) or
+        is_valid_bech32_bech32m_address(value)
+    )
 
 
 def is_valid_bech32_address(value: str) -> bool:
-    """Validates a bitcoin SegWit address for the mainnet
+    """Validates a bitcoin Segwit address using BIP-173
+    https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:49]: Pointing to a particular commit? Yeah sure.
What do you mean by other links?

File: rotkehlchen/chain/bitcoin/utils.py [hunk 6]
```diff
@@ -21,18 +30,90 @@ class OpCodes:
...
+    decoded = bech32.convertbits(data[1:], 5, 8, False)
+    if decoded is None or len(decoded) < 2 or len(decoded) > 40:
+        return False
+    if data[0] > 16:
+        return False
+    if data[0] == 0 and len(decoded) != 20 and len(decoded) != 32:
+        return False
+    if data[0] == 0 and spec != BitcoinSegwitEncoding.BECH32 or data[0] != 0 and spec != BitcoinSegwitEncoding.BECH32M:  # noqa: E501
+        return False
+    return (data[0], decoded) != (None, None)
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:75]: Ikr. I should just return `True`. THanks

File: rotkehlchen/chain/bitcoin/utils.py [hunk 7]
```diff
@@ -21,18 +30,90 @@ class OpCodes:
...
+
+
+def _bech32_verify_checksum(hrp: str, data: List[int]) -> Optional[BitcoinSegwitEncoding]:
+    """Verify a checksum given HRP and converted data characters.
+
+    Code is taken from:
+    https://github.com/sipa/bech32/blob/master/ref/python/segwit_addr.py#L51
+    """
+    const = bech32.bech32_polymod(bech32.bech32_hrp_expand(hrp) + data)
+    if const == 1:
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:110]: Which first case? `bech32.bech32_polymod` returns an int

File: rotkehlchen/chain/bitcoin/utils.py (hunk 1 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:110]: Witver also known as the witness version is simply the version of the address.

BECH32 -> Witness version of `0`
BECH32M -> Witness version of `>=1`

File: rotkehlchen/chain/bitcoin/utils.py [hunk 8]
```diff
@@ -107,21 +188,91 @@ def pubkey_to_p2sh_p2wpkh_address(data: bytes) -> BTCAddress:
...
+    converted_bits = bech32.convertbits(witprog, 8, 5)
+    if converted_bits is None:
+        raise EncodingError('Could not derive Bech32 address from given public key')
+    ret = _bech32_encode([witver] + converted_bits, spec)
+    if _bech32_decode(ret) == (None, None):
+        raise EncodingError('Could not derive Bech32 address from given public key')
+    return BTCAddress(ret)
+
+
+def _tweak_public_key(data: bytes) -> bytes:
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:211]: I don't think that's a common pattern across the codebase. But yeah, you're right

File: rotkehlchen/chain/bitcoin/utils.py [hunk 9]
```diff
@@ -21,18 +30,90 @@ class OpCodes:
...
+    if data[0] > 16:
+        return False
+    if data[0] == 0 and len(decoded) != 20 and len(decoded) != 32:
+        return False
+    if data[0] == 0 and spec != BitcoinSegwitEncoding.BECH32 or data[0] != 0 and spec != BitcoinSegwitEncoding.BECH32M:  # noqa: E501
+        return False
+    return (data[0], decoded) != (None, None)
+
+
+def _bech32_decode(value: str) -> Union[Tuple[str, List[int], BitcoinSegwitEncoding], Tuple[None, None, None]]:  # noqa: E501
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:78]: I think I explained that here https://github.com/rotki/rotki/pull/4714#discussion_r950178503

File: rotkehlchen/chain/bitcoin/utils.py (hunk 2 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:273]: It's tricky to explain as I just got to know about this two days ago but this [comment should help](https://github.com/sipa/bech32/issues/51#issuecomment-554885190).

TLDR: Basically, it's the attempt at rectifying the woes of `BIP-173`

Sorry, you're probably exhausted from hearing BIP BIP BIP 😄

File: rotkehlchen/chain/bitcoin/utils.py (hunk 4 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:120]: Throughout the codebase iirc, only `witver=0` was passed.

The major difference is the `XOR 1` & `XOR 0x2bc830a3` which represents version 0 & > 1 respectively used in `_bech32_create_checksum`

If you see `bech32.bech32_create_checksum`, only `XOR 1`  or `^ 1` is seen.

Comment (nebolax) [reply rotkehlchen/chain/bitcoin/utils.py:120]: Ah, got it, thanks

File: rotkehlchen/chain/bitcoin/utils.py (hunk 2 above)
Comment (nebolax) [reply rotkehlchen/chain/bitcoin/utils.py:273]: Ha-ha, thanks, got it

File: rotkehlchen/chain/bitcoin/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/bitcoin/utils.py:110]: Ah, sorry, my bad. For some reason I thought that `BECH32M_CONST` is a enum value and was wondering why the first comparison is `if const == 1:` while in the second `if const == BECH32M_CONST:` we compare with enum? :smile:

File: rotkehlchen/chain/bitcoin/utils.py [hunk 10]
```diff
@@ -21,18 +30,90 @@ class OpCodes:
...
+    BECH32 = 1  # Native Segwit(P2WPKH)
+    BECH32M = 2  # Taproot(P2TR)
+
+
 def is_valid_btc_address(value: str) -> bool:
-    return is_valid_base58_address(value) or is_valid_bech32_address(value)
+    return (
+        is_valid_base58_address(value) or
+        is_valid_bech32_address(value) or
+        is_valid_bech32_bech32m_address(value)
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/bitcoin/utils.py:43]: Aha, I think I got it. Thank you for the explanation!

(maybe also will make sense to add a bit of this text in the code since I personally was a bit confused when found really little difference between these methods and methods from the `bech32` sourcecode)

File: rotkehlchen/chain/bitcoin/utils.py [hunk 11]
```diff
@@ -21,18 +30,90 @@ class OpCodes:
...
+    if spec is None:
+        return (None, None, None)
+    return (hrp, data[:-6], spec)
+
+
+def _bech32_verify_checksum(hrp: str, data: List[int]) -> Optional[BitcoinSegwitEncoding]:
+    """Verify a checksum given HRP and converted data characters.
+
+    Code is taken from:
+    https://github.com/sipa/bech32/blob/master/ref/python/segwit_addr.py#L51
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py:107]: I don't know about that

File: rotkehlchen/chain/bitcoin/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/bitcoin/utils.py:55]: Oups, didn't replied, sorry. By other links I mean all other github links that you are adding in this PR

File: rotkehlchen/chain/bitcoin/hdkey.py [hunk 12]
```diff
@@ -30,10 +30,11 @@ class PrefixParsingResult(NamedTuple):
     hint: str
 
 
-class XpubType(Enum):
+class XpubType(SerializableEnumMixin):
```
Comment (LefterisJP) [inline rotkehlchen/chain/bitcoin/hdkey.py:33]: use `auto()` here

File: rotkehlchen/chain/bitcoin/hdkey.py [hunk 13]
```diff
@@ -398,6 +406,11 @@ def derive_child(self, idx: Union[int, str]) -> 'HDKey':
         return self._child_from_xpub(index=index, child_xpub=child_xpub)
 
     def address(self) -> BTCAddress:
+        if self.hint == 'xpub' and self.xpub_type == XpubType.P2TR:
+            return pubkey_to_bech32_address(
+                data=self.pubkey.format(COMPRESSED_PUBKEY),
+                witver=1,
```
Comment (LefterisJP) [inline rotkehlchen/chain/bitcoin/hdkey.py:412]: what is this and why is it always constant? If it's a constant `1` then define it somewhere along with an explanation of what it is and use it here

File: rotkehlchen/chain/bitcoin/utils.py [hunk 14]
```diff
@@ -21,18 +29,96 @@ class OpCodes:
     op_checksig = b'\xac'
 
 
+class BitcoinSegwitEncoding(Enum):
+    """Enumeration type to list the various Bitcoin Segwit supported encodings."""
+    BECH32 = 1  # Native Segwit(P2WPKH)
```
Comment (LefterisJP) [inline rotkehlchen/chain/bitcoin/utils.py:34]: use auto

File: rotkehlchen/chain/bitcoin/utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/chain/bitcoin/utils.py:83]: But can you please add some part of that explanation in this docstring because not only Alexey, also I had the same question. And whoever else ends up reading this.

File: rotkehlchen/chain/bitcoin/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/chain/bitcoin/utils.py:115]: Can you define this constant somehow @prettyirrelevant ?

File: rotkehlchen/chain/bitcoin/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/chain/bitcoin/utils.py:219]: Hmm we use libraries where we can. Just copy pasting code is dangerous since it can be full of errors, and we need to maintain it ourselves.

Can't you use an external library for most of this?

**Review by LefterisJP — COMMENTED** [review]
Hmm I think you need to find a library for this. Makes no sense to copy paste code like this for important bitcoin functionality like this. We should not have to maintain this code ourselves. It can't be that there is no python library for bitcoin that offers this?

File: rotkehlchen/chain/bitcoin/utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py]: Okay.

File: rotkehlchen/chain/bitcoin/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/chain/bitcoin/utils.py]: > Hmm I think you need to find a library for this. Makes no sense to copy paste code like this for important bitcoin functionality like this. We should not have to maintain this code ourselves. It can't be that there is no python library for bitcoin that offers this?

Used [bip-utils](https://github.com/ebellocchia/bip_utils). 
Most of its dependencies are already used in the codebase. The only addition are: 
```
cbor2~=5.1
crcmod~=1.7
ecdsa~=0.15
```

**Review by LefterisJP — APPROVED** [review]
LGTM. Thank you Isaac
